### PR TITLE
Do not use thickness specified via PINCH to find uniquepoints.

### PR DESCRIPTION
**Pull request by blattms**
When preprocessing the grid in uniquepoints.c and preprocess.c we used to use the thickness specified via the PINCH keyword when calling preprocess in processEclipseFormat. That seems wrong as it will neglect whether there is MINPV specified at all and whether the cell we remove has a small enough value.

One example is a deck without MINPV, PINCH with thickness t, and with a grid that is a stack cells with a thickness t1 with t1 < t and 2*t1 > t. In that case every second cell would have vanished.

Wondering what would have happend with t bigger than the hight of the grid...

Not sure whether this will break other stuff though.

### CODE REVIEW

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
I am undrafting this as I think that this bugfix is correct. One could argue about always using 1e-18 to find unique points as previously 0 was used if there was no PINCH keyword.

Jenkins seems to disagree. There are two failing test:    
- mpi.compareECLFiles_flow+NORNE_ATW2013_B1H_RE-PERF
```
Keyword: TRANX, origin Init file
Global index (zero based)   = 19332
Grid coordinate             = (10, 78, 10)
(first value, second value) = (45.1909, 45.1889)

Keyword: TRANX requires strict tolerances.

Program threw an exception: [/var/lib/jenkins/workspace/opm-grid-PR-builder/deps/opm-common/test_util/EclRegressionTest.cpp:244] Deviations exceed tolerances.
The absolute deviation is 0.00200653, and the tolerance limit is 1e-06.
The relative deviation is 4.44012e-05, and the tolerance limit is 1e-06.
...
4 keywords exhibit failures
	TRANNNC: Init file
		Fails for 12 entries
		Largest absolute error: 7.9917908e-04
		Largest relative error: 6.0562550e-03
	TRANX: Init file
		Fails for 7 entries
		Largest absolute error: 2.1800995e-03
		Largest relative error: 1.6101299e-04
	TRANY: Init file
		Fails for 8 entries
		Largest absolute error: 3.9978027e-03
		Largest relative error: 5.8321849e-05
	TRANZ: Init file
		Fails for 13 entries
		Largest absolute error: 5.0506592e-03
		Largest relative error: 3.8138157e-05
```
- mpi.compareECLInitFiles_flow+NORNE_ATW2013
```
Comparing TRANX ... 

Keyword: TRANX, origin Init file
Global index (zero based)   = 19332
Grid coordinate             = (10, 78, 10)
(first value, second value) = (45.1889, 45.1909)

Keyword: TRANX requires strict tolerances.

Program threw an exception: [/var/lib/jenkins/workspace/opm-grid-PR-builder/deps/opm-common/test_util/EclRegressionTest.cpp:244] Deviations exceed tolerances.
The absolute deviation is 0.00200653, and the tolerance limit is 1e-06.
The relative deviation is 4.44012e-05, and the tolerance limit is 1e-06.
...
4 keywords exhibit failures
	TRANNNC: Init file
		Fails for 12 entries
		Largest absolute error: 7.9917908e-04
		Largest relative error: 6.0562550e-03
	TRANX: Init file
		Fails for 7 entries
		Largest absolute error: 2.1800995e-03
		Largest relative error: 1.6101299e-04
	TRANY: Init file
		Fails for 8 entries
		Largest absolute error: 3.9978027e-03
		Largest relative error: 5.8321849e-05
	TRANZ: Init file
		Fails for 13 entries
		Largest absolute error: 5.0506592e-03
		Largest relative error: 3.8138157e-05
```

Both models have:
```
PINCH
 0.001 GAP  1* TOPBOT TOP/

MINPV
  500 /
```

Here is my interpretation what master does:
- cells thinner than 0.001 get collapsed irregardless of the pore volume. they are still marked as active.
- Now MinpvProcesor is run:
I would assume that there are cells that previously got removed via `finduniquepoints` (collapses cells thinner than .001 and hence makes them inactive, but it does not seem to change actnum at all). before the grid is processed by `MinpvProcessor`. I assume that due to finduniquepoints these cells have zero volume and are inactive.

**Comment by blattms** [comment]
@tskille This PR fixes your second PINCH problem.

**Comment by blattms** [comment]
I am still unsure whether we should not use 0 (instead of 1e-18) as tolerance for uniquifypoints. Maybe someone clarify what it should do. Is it only there to save space by not storing the equal points multiple times?

**Comment by atgeirr** [comment]
> My assumption is that for some cells a few points did collapse before and with this PR do not anymore. The cells where this happens now are less degenerated and have different volume. This will also result in different transmissibilities.
> 
> The `MinPVProcessor` operates on information of the EclGrid representation only, and hence should still do the same thing.
> 
> Dose that make sense?

I think it does. Since we are seeking to emulate Eclipse behaviour when it comes to keywords like PINCH, I guess comparing the fields in question to the Eclipse output of the same should settle the issue.



> I am still unsure whether we should not use 0 (instead of 1e-18) as tolerance for uniquifypoints. Maybe someone clarify what it should do. Is it only there to save space by not storing the equal points multiple times?

For a cartesian grid, 8 cells meet at a single vertex. In a corner point grid, the 8 cells that meet at a vertex in a logical cartesian sense may all have different coordinates for that "vertex", however usually it is not quite that bad. The question of unique points here is one of identifying the topology. We must identify the points that are identical to make the (purely integer-based) building of the topology work. I think 0 should be used rather than 1e-18. The processing of each point should give the same result if the input is the same, so there should be no need for a tolerance there.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
benchmark please

**Comment by atgeirr** [comment]
> removed the tolerance parameter from processEclipseFormat

I think there may be calls in the upscaling repo that calls the same, you may see some fallout. I do not see any harm in simply keeping it?

**Comment by blattms** [comment]
Nope:
```
mblatt@smaug:~/src/dune/opm/opm-upscaling$ find opm/ -type f | xargs grep processEclipseFormat
opm/porsol/common/setupGridAndProps.hpp:                grid.processEclipseFormat(&inputGrid, nullptr, periodic_extension, turn_normals, clip_z);
opm/porsol/common/setupGridAndProps.hpp:        grid.processEclipseFormat(&eg, nullptr, periodic_extension, turn_normals, clip_z);
opm/upscaling/initCPGrid.cpp:        grid.processEclipseFormat(&inputGrid, nullptr, periodic_extension , turn_normals );
```
The documentation of the method even says that it is internal. I might try to make it private or remove it from CpGrid altogether in another PR.

Removing an ignored parameter seems cleaner/clearer (I still keep in uniquifypoints, though).

**Comment by atgeirr** [comment]
> Removing an ignored parameter seems cleaner/clearer (I still keep in uniquifypoints, though).

Agree.

**Review by atgeirr — APPROVED** [review]

**Comment by ytelses** [comment]
Benchmark result overview: 




| Test    | Configuration                                     | Relative |
| ------- | ------------------------------------------------- | -------- |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 1        | 1.011    |
| opm-git | OPM Benchmark: flow_mpi_extra - Threads: 8        | 1        |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 1        | 0.986    |
| opm-git | OPM Benchmark: flow_mpi_norne - Threads: 8        | 1.046    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 1 | 1.004    |
| opm-git | OPM Benchmark: flow_mpi_norne_4c_msw - Threads: 8 | 1.165    |
* Speed-up = Total time master / Total time pull request. Above 1.0 is an improvement. *



View result details @ https://www.ytelses.com/opm/?page=result&id=1213

**Comment by blattms** [comment]
[well plots for current master and this fix](https://github.com/OPM/opm-grid/files/6569467/norne-wells.pdf) look reasonable to me.

**Comment by blattms** [comment]
mmh, https://github.com/OPM/opm-tests/pulls gives a HTTP 500 error. should I issue an update_data. anyway?

**Comment by blattms** [comment]
Sigh, I guess there is no opencl platform on the jenkins system:
```
Running 1 test case...
%%MatrixMarket
%%MatrixMarket
unknown location(0): [NON-XML-CHAR-0x1B][4;31;49mfatal error: in "TestDefaultPreconditionerFactory": std::logic_error: [/var/lib/jenkins/workspace/opm-simulators-PR-builder@2/opm/simulators/linalg/bda/openclSolverBackend.cpp:182] OpenCL Error: clGetPlatformIDs(-1001)
UNKNOWN_CL_CODE[NON-XML-CHAR-0x1B][0;39;49m
/var/lib/jenkins/workspace/opm-simulators-PR-builder@2/tests/test_openclSolver.cpp(104): [NON-XML-CHAR-0x1B][1;36;49mlast checkpoint: "TestDefaultPreconditionerFactory" test entry[NON-XML-CHAR-0x1B][0;39;49m

[NON-XML-CHAR-0x1B][1;31;49m*** 1 failure is detected in the test module "OPM_test_openclSolver"
[NON-XML-CHAR-0x1B][0;39;49m--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpiexec detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[65344,1],0]
  Exit code:    201
```
I guess you will have to work around this one...

**Comment by GitPaean** [comment]
we could not access `https://github.com/OPM/opm-tests/pulls ` for long. We can only access specific pull request directly with link.

**Comment by bska** [comment]
> we could not access https://github.com/OPM/opm-tests/pulls for long.

True.  There **is** a workaround, but that requires using your phone.  I installed the GitHub app on my phone a while back, and I'm able to view the active PRs in opm-tests in the app.

**Comment by bska** [comment]
> I'm able to view the active PRs in opm-tests in the app.

For the record, here are the **currently** open (2021-06-07) PRs in OPM-tests

* OPM/opm-tests#191
* OPM/opm-tests#367
* OPM/opm-tests#396
* OPM/opm-tests#525

**Comment by GitPaean** [comment]
>  I'm able to view the active PRs in opm-tests in the app.

Good to know. At least there is a way that we can know the list of the open PRs there.

**Comment by blattms** [comment]
Thanks @bska. These all seem rather stale. Hence there is nothing important and I will go ahead with this one.

**Comment by blattms** [comment]
jenkins build this with downstreams update_data please

**Comment by bska** [comment]
> These all seem rather stale.

Agreed, with the exception of OPM/opm-tests#367.  I'd like to get that one merged Soon[tm].

**Comment by blattms** [comment]
Shall I cancel the build then?

**Comment by blattms** [comment]
Just checked. I do not assume the my build will change any of those files. Hence no merge conflicts should appear.

Anywyay, I can wait until the other is merged.

**Comment by bska** [comment]
> Shall I cancel the build then?

No, there's no need.  The PR I referenced "only" adds a few new simulation models (i.e., `.DATA` files).  The only reason it's not merged is that it also adds new summary vectors to an include file that's used by a number of existing regression tests and there's a little bit of chicken-and-egg juggling needed to get the updates right in this case.  The principle blocker is that we can't trigger a data update from opm-tests itself.

> Anyway, I can wait until the other is merged.

No, there's no reason to do that.  Please go ahead with your merge.  I'll deal with OPM/opm-tests#367.

**Comment by blattms** [comment]
problems pushing to opm-tests, Maybe I am doing something wrong @akva2?

```
To github.com:jenkins4opm/opm-tests
 ! [remote rejected] master (refusing to delete the current branch: refs/heads/master)
error: failed to push some refs to 'git@github.com:jenkins4opm/opm-tests'
To github.com:jenkins4opm/opm-tests
 - [deleted]         update_opm-simulators_3115
... using saved access token
created issue 548
pushing branch update_opm-grid_538 to jenkins4opm
error: src refspec update_opm-grid_538 does not match any.
error: failed to push some refs to 'git@github.com:jenkins4opm/opm-tests'
Build step 'Execute shell' marked build as failure
Archiving artifacts
Recording test results
ERROR: Step ?Publish JUnit test result report? failed: No test report files were found. Configuration error?
[GNU C Compiler (gcc)] Skipping execution of recorder since overall result is 'FAILURE'
[Checks API] No suitable checks publisher found.
Setting status of d10a812b2f261dd3f9daeea1bf066b43906b41c7 to FAILURE with url https://ci.opm-project.org/job/opm-grid-PR-builder/216/ and message: 'Build finished. '
```

**Comment by akva2** [comment]
see further up, pip issues, no idea what's going on.

**Comment by bska** [comment]
> pip issues, no idea what's going on.

**Possibly** related to the global outage of cloud service provider "fastly" that's being reported right now.

**Comment by bska** [comment]
jenkins build this update_data please

**Comment by blattms** [comment]
maybe we need to add setuptools-scm somewhere like [here](https://github.com/zephyrproject-rtos/west/commit/70b6892575c5d209f12d382a8b2f4deeb7bbd749)? But it seems like opm-simulators does not even have requirements.txt.

But maybe it is just  a proxy issue. Let's see how the new build performs.

**Comment by akva2** [comment]
problem happened while building the opm-common upstream. but it was indeed a mirror issue, already got a couple of builds past that state again..

**Comment by blattms** [comment]
jenkins build this opm-tests=549 please

**Comment by bska** [comment]
> jenkins build this opm-tests=549 please

May not get to the finish line.  The [previous build (217)](https://ci.opm-project.org/job/opm-grid-PR-builder/217/console) failed to compile (in opm-upscaling).

**Comment by bska** [comment]
New reference solutions have been installed on the CI system.  Merging this now to bring the overall system back to a consistent state.

**Comment by blattms** [comment]
Thanks for merging and the valuable comments.

**Pull request merged**
